### PR TITLE
[Impeller] fixed Rect::Contains

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/140464

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/geometry/rect.h [hunk 1]
```diff
@@ -130,7 +130,8 @@ struct TRect {
   }
 
   [[nodiscard]] constexpr bool Contains(const TRect& o) const {
```
Comment (gaaclarke) [inline impeller/geometry/rect.h:132]: We want `a.Contains(a) == true`, right?

Comment (bdero) [reply impeller/geometry/rect.h:132]: Yes indeed, we assume this inclusivity in our current usages

**Review by bdero — APPROVED** [review]
LGTM. This looks safe because the Get[Direction] methods return sorted directions.

File: impeller/geometry/rect.h (hunk 1 above)
Comment (gaaclarke) [reply impeller/geometry/rect.h:132]: I only ask because the Contains(Point) uses < operators for the right and bottom axises.

Comment (flar) [reply impeller/geometry/rect.h:132]: Another issue with these methods is that empty can contain another rectangle and vice versa. I'm planning to address all of these in a rewrite now that the fields are hidden, but I'm just starting on that now and if this PR fixes an existing practical issue, then we should proceed with it.

**Comment by flar** [comment]
> LGTM. This looks safe because the Get[Direction] methods return sorted directions.

Actually, the opposite. Because of the sorted data an empty rectangle can "contain" a non-empty rectangle. The auto-sorting of rectangles is one of the first things I'm going to fix in the rework of Rect.

**Comment by flar** [comment]
The eventual implementation of these methods will be based upon the versions that were created for the DL-specific geometry classes, such as:

https://github.com/flar/engine/blob/f391c31b882278603f42b725b3587fc94dddc0ab/display_list/geometry/dl_rect.h#L553

Note that we decided to not proceed with the DL-geometry package and instead have it share a geometry package with Impeller by updating Impeller's classes.

File: impeller/geometry/rect.h (hunk 1 above)
Comment (flar) [reply impeller/geometry/rect.h:132]: The left and top edges are considered "inside", the right and bottom edges are considered "outside".

For points you want `>= left/top` and `< right/bottom` to match that rule.

For rects, both have the same insideness and so matching edges always have the same containment as each other. In the case of right and bottom, those edges are the non-inclusive limit of what each rectangle considers "inside" and so if they are equal then all of the points inside one of them are also inside the other (wrt that one edge).

**Pull request merged**